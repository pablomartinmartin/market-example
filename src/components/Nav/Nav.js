import React, { useState, useEffect } from "react";
import { withRouter } from "react-router-dom";

import logo from "../../assets/Logo_ML.png";
import search from "../../assets/ic_Search.png";

import "./Nav.scss";

const Nav = (props) => {
  const [query, setQuery] = useState('');

  const queryParams = new URLSearchParams(props.location.search);
  const querySearch = queryParams.get('search');

  useEffect(() => {
    if (querySearch) {
      setQuery(querySearch);
    }
  }, [querySearch]);

  const handleSubmitSearch = (event) => {
    event.preventDefault();

    props.history.push({
      pathname: '/items',
      search: `?search=${query}`,
    });
  };

  return (
    <nav className="nav-search">
      <img src={logo} alt="Logo" className="nav-search__logo" />
      <form name="search-form" onSubmit={handleSubmitSearch}>
        <input
          type="text"
          name="searchField"
          placeholder="Nunca dejes de buscar"
          maxLength="100"
          autoFocus
          className="nav-search__field"
          onChange={(e) => setQuery(e.target.value)}
          defaultValue={query || ""}
        />
        <button type="submit" className="nav-search__button">
          <img src={search} alt="Search" />
        </button>
      </form>
    </nav>
  );
};

export default withRouter(Nav);
