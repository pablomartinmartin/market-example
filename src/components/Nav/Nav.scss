@import "../../App.scss";

$search-height: 35px;
$search-font-size: 18px;

.nav-search {
  display: flex;
  height: 14vh;
  background-color: $ml-yellow;
  justify-content: center;
  align-items: center;
  

  > form {
    display: flex;
    width: 70%;
  }

  &__logo {
    display: flex;
    margin: 0 20px;
  }

  &__field {
    color: $ml-black;
    font-size: $search-font-size;
    height: $search-height;
    padding: 5px 10px;
    width: 90%;
    border: 0;
    border-radius: 3px 0 0 3px;
  }

  &__field::placeholder {
    color: $ml-gray;
    font-size: $search-font-size;
  }

  &__button {
    background-color: $ml-light-gray;
    border: 0;
    border-left: 0;
    border-radius: 0 3px 3px 0;
    height: $search-height + 10px;
    padding: 5px 10px;
  }
}
