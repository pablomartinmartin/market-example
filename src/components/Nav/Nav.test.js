import React from 'react';
import { mount } from 'enzyme';
import Nav from './Nav';

const history = {
  push: jest.fn(),
};

const mountComponent = (props) => {
  const component = mount(
        <Nav.WrappedComponent history={history} {...props} />
  );

  return component;
};

describe('<Nav />', () => {
  let component;

  beforeEach(() => {
    component = mountComponent();
  });

  it('renders nav component', () => {
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  describe('Search actions', () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation((init) => [init, setState]);

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('renders search button', () => {
      expect(component.find('button')).toBeDefined();
    });    
    
    it('renders search input', () => {
      expect(component.find('input')).toBeDefined();
    });

    it('should call the search function', () => {
      const component = mountComponent();

      const searchButton = component.find('button');

      searchButton.simulate('submit');
      component.update();

      expect(history.push).toHaveBeenCalled();
      expect(history.push).toHaveBeenCalledWith({
        pathname: '/items',
        search: `?q=`,
      });
    });

    it('should send the query', () => {
      const component = mountComponent();
      const searchInput = component.find('input');

      searchInput.simulate('change', {target: {name: 'searchField', value: 'sillon'}});
      searchInput.simulate('submit');
      component.update();

      expect(history.push).toHaveBeenCalled();
      expect(history.push).toHaveBeenCalledWith({
        pathname: '/items',
        search: `?q=sillon`,
      });
    });
  });
});
