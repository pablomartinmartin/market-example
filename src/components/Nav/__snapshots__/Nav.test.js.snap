// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders nav component 1`] = `
<Nav
  history={
    Object {
      "push": [MockFunction],
    }
  }
>
  <nav
    className="nav-search"
  >
    <img
      alt="Logo"
      className="nav-search__logo"
      src="Logo_ML.png"
    />
    <form
      name="search-form"
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        className="nav-search__field"
        maxLength="100"
        name="searchField"
        onChange={[Function]}
        placeholder="Nunca dejes de buscar"
        type="text"
      />
      <button
        className="nav-search__button"
        type="submit"
      >
        <img
          alt="Search"
          src="ic_Search.png"
        />
      </button>
    </form>
  </nav>
</Nav>
`;
